npm_and_yarn/lib/dependabot/npm_and_yarn/version.rb

```diff
@@ -34,6 +34,27 @@ def initialize(version)
         super
       end
 
+      def major
+        @major ||= segments[0] || 0
+      end
+
+      def minor
+        @minor ||= segments[1] || 0
+      end
+
+      def patch
+        @patch ||= segments[2] || 0
+      end
+
+      def backwards_compatible_with?(other)
```

Comment:
My 🪙🪙 -- I think this method name is a little misleading _on this object_. I understand this to be specific to `@types` packages, where we're treating `@types/foo@3.5` as compatible with `foo@3.5`. However, I don't think this is generalizable to other use cases.

Since this method is on `NpmAndYarn::Version`, I'd like to see us adopt a name that _is_ generalizable and true in all cases. I'll try to think of some names and leave them here in a follow-up comment 🙂

Reply:
My thinking here is that the breaking changes in SemVer refer to the package _itself_, while this is (at least how we're using it so far) comparing two distinct packages. Does that make sense @pavera @mattt 

Maybe something like `shares_stable_version_with?`? or `eqivalent_stable_version?`

Reply:
As I understand it, this `Version` object represents only the version of a package, independent of any particular package release. Given two releases of the same package, this method _should_ provide the correct answer for whether they're backwards compatible — at least as I understand how that's defined in the [SemVer 2 spec](https://semver.org/)

> Given a version number MAJOR.MINOR.PATCH, increment the:
>
> * MAJOR version when you make incompatible API changes,
> * MINOR version when you add functionality in a backwards compatible manner, and
> * PATCH version when you make backwards compatible bug fixes

So we should be able to use `backwards_compatible_with?` to compare versions of either the same package or a `@types` package and its corresponding package. The fact that this corresponds with how `@types` packages are associated is a happy coincidence.

Reply:
> Do you see any particular test case that you find surprising / breaks your expectation of how this method should behave?

I don't. It's odd to me that our fist use case is comparing versions of separate packages. Your reply highlights that I don't have a good understanding of my own reservations. Since I don't have a better way to say it, let's move forward with what we have 😄

Reply:
> My thinking here is that the breaking changes in SemVer refer to the package _itself_, while this is (at least how we're using it so far) comparing two distinct packages. Does that make sense @pavera @mattt
> 
> Maybe something like `shares_stable_version_with?`? or `eqivalent_stable_version?`

(Sorry, my last comment came right after yours)

OK, that really helped me understand your concern, I think. 

It sounds to me that your objection is more with how this method is used than its name — that we're saying "3.5.0" is backwards compatible with "3.6.1", even though we're talking about two different packages. Would you agree with that?

If we change the name, we'd still be comparing 🍎s to 🟠s. It just happens that, in the case of `@types` packages, it's 🍎 to 🍏 , so it works out.

We _could_ raise up this API to the level of comparing a tuple of `(Package, Version)`, which could tell you that you're comparing 🍎s to 🟠s. But for the one place that we're making this inter-package comparison, I'm personally alright with leaving things as they are.